packages/js/product-editor/src/blocks/text/edit.tsx

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { createElement, createInterpolateElement } from '@wordpress/element';
+import { useBlockProps } from '@wordpress/block-editor';
+import { useInstanceId } from '@wordpress/compose';
+import { __, sprintf } from '@wordpress/i18n';
+import { Product } from '@woocommerce/data';
+import classNames from 'classnames';
+import {
+	BaseControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { useValidation } from '../../contexts/validation-context';
+import useProductEntityProp from '../../hooks/use-product-entity-prop';
+import { TextBlockAttributes } from './types';
+
+export function Edit( { attributes }: { attributes: TextBlockAttributes } ) {
+	const blockProps = useBlockProps();
+	const {
+		property,
+		label,
+		placeholder,
+		required,
+		validationRegex,
+		validationErrorMessage,
+		minLength,
+		maxLength,
+	} = attributes;
+	const [ value, setValue ] = useProductEntityProp( property );
+	const nameControlId = useInstanceId( BaseControl, property ) as string;
+
+	const { error, validate } = useValidation< Product >(
+		property,
+		async function validator() {
+			if ( required && ! value ) {
+				return __( 'This field is required.', 'woocommerce' );
+			}
+			if ( validationRegex ) {
+				const regExp = new RegExp( validationRegex );
+				if ( ! regExp.test( value ) ) {
+					return (
+						validationErrorMessage ||
+						__( 'Invalid value for the field.', 'woocommerce' )
+					);
+				}
+			}
+			if ( typeof minLength === 'number' && value.length < minLength ) {
+				return sprintf(
+					/* translators: %d: minimum length */
+					__(
+						'The minimum length of the field is %d',
+						'woocommerce'
+					),
+					minLength
+				);
+			}
+			if ( typeof maxLength === 'number' && value.length > maxLength ) {
+				return sprintf(
+					/* translators: %d: maximum length */
+					__(
+						'The maximum length of the field is %d',
+						'woocommerce'
+					),
+					maxLength
+				);
+			}
+		},
+		[ value ]
+	);
+
+	return (
+		<div { ...blockProps }>
+			<BaseControl
+				id={ nameControlId }
+				label={
+					required
+						? createInterpolateElement( `${ label } <required/>`, {
+								required: (
+									<span className="woocommerce-product-form__required-input">
+										{ /* translators: field 'required' indicator */ }
+										{ __( '*', 'woocommerce' ) }
+									</span>
+								),
+						  } )
```

Comment:
I'm okay with this being in here for this PR, but looking forward, as we create more generic blocks, and look at our existing blocks, we should see if there is a way to implement the label in a consistent manner so that each block does not have to duplicate this sort of logic and formatting.

Reply:
I agree. Ideally, I think BaseControl itself should support this. As a workaround, we can create a wrapper to BaseControl with a `required` prop that adds this.